core/executor/src/wasm_runtimes_cache.rs

```diff
@@ -0,0 +1,328 @@
+	/// sandboxed Wasm runtime.
+	///
+	/// `ext` - Externalities to use for the runtime. This is used for setting
+	/// up an initial runtime instance. The parameter is only needed for calling
+	/// into the Wasm module to find out the `Core_version`.
+	///
+	/// `default_heap_pages` - Number of 64KB pages to allocate for Wasm execution.
+	/// Defaults to `DEFAULT_HEAP_PAGES` if `None` is provided.
+	///
+	/// # Return value
+	///
+	/// If no error occurred a tuple `(wasmi::ModuleRef, Option<RuntimeVersion>)` is
+	/// returned. `RuntimeVersion` is contained if the call to `Core_version` returned
+	/// a version.
+	///
+	/// In case of failure one of two errors can be returned:
+	///
+	/// `Err::InvalidCode` is returned for runtime code issues.
+	///
+	/// `Error::InvalidMemoryReference` is returned if no memory export with the
+	/// identifier `memory` can be found in the runtime.
+	pub fn fetch_runtime<E: Externalities<Blake2Hasher>>(
+		&mut self,
+		wasm_executor: &WasmExecutor,
+		ext: &mut E,
+		default_heap_pages: Option<u64>,
+	) -> Result<Rc<CachedRuntime>, Error> {
+		let code_hash = ext
+			.original_storage_hash(well_known_keys::CODE)
+			.ok_or(Error::InvalidCode)?;
+
+		// This is direct result from fighting with borrowck.
+		let handle_result =
+			|cached_result: &Result<Rc<CachedRuntime>, CacheError>| match *cached_result {
+				Err(_) => Err(Error::InvalidCode),
+				Ok(ref cached_runtime) => Ok(Rc::clone(cached_runtime)),
+			};
+
+		match self.instances.entry(code_hash.into()) {
+			Entry::Occupied(o) => handle_result(o.get()),
+			Entry::Vacant(v) => {
+				trace!(target: "runtimes_cache", "no instance found in cache, creating now.");
+				let result = Self::create_wasm_instance(wasm_executor, ext, default_heap_pages);
+				if let Err(ref err) = result {
+					warn!(target: "runtimes_cache", "cannot create a runtime: {:?}", err);
+				}
+				handle_result(v.insert(result))
+			}
+		}
+	}
+
+	fn create_wasm_instance<E: Externalities<Blake2Hasher>>(
+		wasm_executor: &WasmExecutor,
+		ext: &mut E,
+		default_heap_pages: Option<u64>,
+	) -> Result<Rc<CachedRuntime>, CacheError> {
+		let code = ext
+			.original_storage(well_known_keys::CODE)
+			.ok_or(CacheError::CodeNotFound)?;
+		let module = WasmModule::from_buffer(&code).map_err(|_| CacheError::InvalidModule)?;
+
+		// Extract the data segments from the wasm code.
+		//
+		// A return of this error actually indicates that there is a problem in logic, since
+		// we just loaded and validated the `module` above.
+		let data_segments = extract_data_segments(&code).ok_or(CacheError::CantDeserializeWasm)?;
+
+		let heap_pages = ext
+			.storage(well_known_keys::HEAP_PAGES)
+			.and_then(|pages| u64::decode(&mut &pages[..]))
+			.or(default_heap_pages)
+			.unwrap_or(DEFAULT_HEAP_PAGES);
+
+		// Instantiate this module.
+		let instance = WasmExecutor::instantiate_module::<E>(heap_pages as usize, &module)
+			.map_err(|_| CacheError::Instantiation)?;
```

Comment:
This error shouldn't get silently dropped -- it made it harder to find the issue with `__heap_base`.